### PR TITLE
ConditionVariable: add predicate wait methods

### PR DESCRIPTION
**Pull request by lrusak**
This is the second part of my locking/conditions/events cleanup.

This adds support for the `std::condition_variable_any::wait` methods that allow specifying a predicate. We did this before with the class `TightConditionVariable` which allowed specifying a templated predicate.

This PR changes the predicate to use `std::function<bool()>`. This is then satisfied with various conditions and lambdas.

I hope that the stl method works similar enough to our old method (our tests pass at least).

It would be nice to remove some of these passthrough methods but it seems that there is a conflict with the CSingleExit class and the lock count. Looking at this old commit from @elupus https://github.com/xbmc/xbmc/commit/df90dddce134a426e7dcb56e1c2cd057fb2f6fe5 it seems that this is needed to keep track of the locking count in some cases. I'll try and look into this more and possibly write a test case so we can confirm the behaviour (and possibly improve it). Maybe @jimfcarroll has some ideas here.

This PR allows removing of the `TightConditionVariable` and `InversePredicate` classes.

### CODE REVIEW

File: xbmc/threads/Condition.h [hunk 1]
```diff
@@ -38,6 +47,18 @@ namespace XbmcThreads
       lock.count = count;
     }
 
+    template<typename Rep, typename Period>
+    inline bool wait(CCriticalSection& lock,
+                     std::chrono::duration<Rep, Period> duration,
+                     std::function<bool()> predicate)
+    {
+      int count = lock.count;
+      lock.count = 0;
```
Comment (garbear) [inline xbmc/threads/Condition.h:56]:
```c++
int count = 0;
std::swap(count, lock.count);
```

File: xbmc/threads/Condition.h [hunk 2]
```diff
@@ -48,8 +69,21 @@ namespace XbmcThreads
...
     }
 
+    inline void wait(CSingleLock& lock, std::function<bool()> predicate)
+    {
+      cond.wait(lock.get_underlying(), predicate);
+    }
+
     inline void wait(CSingleLock& lock) { wait(lock.get_underlying()); }
 
+    template<typename Rep, typename Period>
```
Comment (garbear) [inline xbmc/threads/Condition.h:79]: `Rep` stands for Representation?

Comment (lrusak) [reply xbmc/threads/Condition.h:79]: Yep! I tried to use the same nomenclature as the official templates.
```
rep	Rep, an arithmetic type representing the number of ticks
period	Period, a [std::ratio] representing the tick period (i.e. the number of second's fractions per tick)
```
ref: https://en.cppreference.com/w/cpp/chrono/duration

File: xbmc/threads/Condition.h (hunk 1 above)
Comment (lrusak) [reply xbmc/threads/Condition.h:56]: interesting, I'm hesitant to deviate from the predefined pattern though.

**Review by garbear — APPROVED** [review]

**Comment by lrusak** [comment]
jenkins build this please

**Pull request merged**